### PR TITLE
Fix/30868: Incorrect group displayed

### PR DESCRIPTION
**Pull request by DylanDylann**
### Details

- Web - Chat- A different group chat appears when refreshing the page
### Fixed Issues

$ https://github.com/Expensify/App/issues/30868
PROPOSAL: https://github.com/Expensify/App/issues/30868#issuecomment-1793304559

### Tests

1. Create a group conversation
2. Type something
3. Clicking on a profile
4. Refresh the page
5. The same group chat appears when refreshing the page
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Create a group conversation
2. Type something
3. Clicking on a profile
4. Refresh the page
5. The same group chat appears when refreshing the page
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

- Cannot test in Android Native
</details>

<details>
<summary>Android: mWeb Chrome</summary>

[Screencast from 23-11-2023 11:09:57.webm](https://github.com/Expensify/App/assets/141406735/17d8804f-a521-4413-adb6-b0f644d14213)

</details>

<details>
<summary>iOS: Native</summary>

- Cannot test in IOS native
</details>

<details>
<summary>iOS: mWeb Safari</summary>

[Screencast from 23-11-2023 11:14:36.webm](https://github.com/Expensify/App/assets/141406735/ca432c1b-f486-42ef-8d32-1db6ee1ccc68)

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

[Screencast from 23-11-2023 11:13:18.webm](https://github.com/Expensify/App/assets/141406735/fa65198b-9402-4f5b-8110-2e55b5d9c531)

</details>

<details>
<summary>MacOS: Desktop</summary>

[Screencast from 23-11-2023 19:01:40.webm](https://github.com/Expensify/App/assets/141406735/ec5ef34a-6700-4b6f-b6c5-6037b2877ddd)

</details>

### CODE REVIEW

**Comment by DylanDylann** [comment]
> @DylanDylann Can you post a video demo for the issue?

Sure. Please wait a minute

File: src/pages/home/ReportScreen.js [hunk 1]
```diff
@@ -160,6 +160,14 @@ function ReportScreen({
...
 
     const reportID = getReportID(route);
+
+    useEffect(() => {
+        if (!reportID) {
+            return;
+        }
+        Report.updateLastVisitTime(reportID);
+    }, [reportID]);
+
```
Comment (DylanDylann) [inline src/pages/home/ReportScreen.js:170]: @techievivek updated to just update the ```lastVisitTime``` if the reportID changes

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: Is there any specific reason to not put lastVisitTime in OpenReport (maybe in optimistic onyx update)? I am thinking only offline behaviour needs to be checked in this.

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: @shubham1206agra There are a few cases we do not call the ```OpenReport``` API

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: Can you point out the cases?

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: https://github.com/Expensify/App/blob/f52c59d55be2d6d2775623d92c3043bd6d42cc0f/src/pages/home/report/ReportActionsView.js#L105-L112

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: Pls help check the above case

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: But in this, don't we have `lastReadTime` as the latest?

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: Before, I tried the solution you`ve mentioned " put lastVisitTime in OpenReport", and when testing, I encountered a few cases that the OpenReport is not called. So we need to make sure the ```lastVisitTime``` is always updated once we open report

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: Can you tell me more about these cases?

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: For example, we create new report in offline mode

Comment (getusha) [reply src/pages/home/ReportScreen.js:170]: Is there any reason to not use the stack navigator listener?

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: In stack navigation, the ```Report.updateLastVisitTime(reportID);``` is called once the navigation stack changes, even if the user just navigates to other screens like setting, profile, ...

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: > For example, we create new group in offline mode, so the field `props.report.isOptimisticReport` is true

But here we should have `lastReadTime` here

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: @shubham1206agra What do you mean?

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: https://github.com/Expensify/App/pull/31424#discussion_r1403927078

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js:170]: @shubham1206agra There is the case:
1. Force offline mode
2. Create group A (report A)
3. Open report B and send message
4. Open report A again (The ```lastVisitTime``` is not updated if follow this [solution](https://github.com/Expensify/App/pull/31424#discussion_r1403923833)

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: @DylanDylann, I am actually not a fan of the current solution because if we try to open an invalid report, then it should not open any report, but it will create a report under `report_` collection, which is undesired in any case, and may cause errors.

Comment (shubham1206agra) [reply src/pages/home/ReportScreen.js:170]: the use effect maybe not triggered when pressing browser back button as the previous ReportScreen is cached and not updated

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: @shubham1206agra Updated the PR to fix your comments:
1. Does not update ```lastVisitTime``` with the empty report
2. Update the ```lastVisitTime``` when going back

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: Please help check

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: @shubham1206agra Please help review once you have a chance

**Comment by shubham1206agra** [comment]
Hey
Can you confirm this problem?
When I go back, openReport is triggered again.
Doesn't seems to be happening on main

File: src/pages/home/ReportScreen.js (hunk 1 above)
Comment (getusha) [reply src/pages/home/ReportScreen.js]: I recommend utilizing the Navigation state event and filtering out states other than Report.

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: @getusha It leads to update ```lastVisitTime``` with the empty report

Comment (getusha) [reply src/pages/home/ReportScreen.js]: @DylanDylann why can't we just filter empty reports? or am i missing something?

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: @getusha Maybe cannot. Currently, when we open the invalid report, it still call API to get this report. Then, if the returned report does not have ```report.reportID```, then we know it is an invalid report and display "Not found view"

**Comment by DylanDylann** [comment]
@shubham1206agra Just checked, I see that in staging, the API 'OpenReport' still called when clicking on "Back". 
[Screencast from 30-11-2023 02:15:30.webm](https://github.com/Expensify/App/assets/141406735/e2a0ea8a-6653-42ac-8988-4f5b6ce8ff02)

File: src/pages/home/ReportScreen.js [hunk 2]
```diff
@@ -164,6 +165,15 @@ function ReportScreen({
...
     const reportID = getReportID(route);
+
+    const isFocused = useIsFocused();
+    useEffect(() => {
+        if (!report.reportID || !isFocused) {
+            return;
+        }
+        Report.updateLastVisitTime(report.reportID);
+    }, [report.reportID, isFocused]);
+
```
Comment (shubham1206agra) [inline src/pages/home/ReportScreen.js:176]: Can you add comment here on why are we doing this?

File: src/pages/home/ReportScreen.js [hunk 3]
```diff
@@ -167,7 +167,10 @@ function ReportScreen({
     const reportID = getReportID(route);
 
     const isFocused = useIsFocused();
+
+    // Update the lastVisitTime when the report is focused
```
Comment (getusha) [inline src/pages/home/ReportScreen.js:171]: > We need to update the `lastVisitTime` every time the report is focused because `lastReadTime` does not have up-to-date time stamp every time the user visits a report.

Lets update it to this.

File: src/pages/home/ReportScreen.js [hunk 4]
```diff
@@ -167,7 +167,10 @@ function ReportScreen({
...
     useEffect(() => {
+        // We don't want to update the lastVisitTime if the report did not exist or if the report is not focused
```
Comment (getusha) [inline src/pages/home/ReportScreen.js:173]: We don't need this one since the code explains itself.

File: src/pages/home/ReportScreen.js (hunk 3 above)
Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: Updated

Comment (DylanDylann) [reply src/pages/home/ReportScreen.js]: Please help check again @getusha

**Review by shubham1206agra — APPROVED** [review]

**Comment by shubham1206agra** [comment]
@getusha Do you also want to soft review this?

**Comment by MonilBhavsar** [comment]
Merging main should pass the performance tests

File: src/libs/actions/Report.js [hunk 5]
```diff
@@ -2502,6 +2502,17 @@ function searchInServer(searchInput) {
     debouncedSearchInServer(searchInput);
 }
 
+/**
+ * @param {String} reportID
+ */
+
```
Comment (MonilBhavsar) [inline src/libs/actions/Report.js:2508]: Unnecessary line
```suggestion
```

File: src/pages/home/ReportScreen.js [hunk 6]
```diff
@@ -165,6 +166,18 @@ function ReportScreen({
     const [scrollPosition, setScrollPosition] = useState({});
 
     const reportID = getReportID(route);
+
+    const isFocused = useIsFocused();
+
+    // We need to update the lastVisitTime every time the report is focused
+    // because lastReadTime does not have up-to-date time stamp every time the user visits a report.
```
Comment (MonilBhavsar) [inline src/pages/home/ReportScreen.js:173]:
```suggestion
    // because lastReadTime is not updated every time the user visits a report.
```

File: src/types/onyx/Report.ts [hunk 7]
```diff
@@ -37,6 +37,9 @@ type Report = {
     /** The last time the report was visited */
```
Comment (MonilBhavsar) [inline src/types/onyx/Report.ts:37]: Can we please update the comment here - 
My suggestion:
`The time when user read the last message`

File: src/libs/ReportUtils.ts [hunk 8]
```diff
@@ -559,10 +559,10 @@ function isDraftExpenseReport(report: OnyxEntry<Report>): boolean {
  */
 function sortReportsByLastRead(reports: OnyxCollection<Report>): Array<OnyxEntry<Report>> {
```
Comment (MonilBhavsar) [inline src/libs/ReportUtils.ts:560]: Looks like this function is only used in `findLastAccessedReport()` right?

**Review by MonilBhavsar — COMMENTED** [review]
Looks good, minor comments

**Comment by shubham1206agra** [comment]
@DylanDylann Can you check performance regression here?

**Comment by DylanDylann** [comment]
@shubham1206agra @MonilBhavsar Do you have any idea about this Reasure Performance Tests issue: 
image

**Comment by shubham1206agra** [comment]
@DylanDylann Can you try to use a new collection for this in another PR just to confirm the performance hit?

**Comment by DylanDylann** [comment]
@shubham1206agra Thanks. I also plan to try it

**Comment by DylanDylann** [comment]
> we want to update lastVisitTime when a report is initially rendered

@MonilBhavsar I don`t think so. We just update the ```lastVisitTime``` if the report.reportID is not empty to prevent it from updating the invalid report. In the initially rendered, there is the case report.reportID is empty

**Comment by getusha** [comment]
@DylanDylann I think it is because of the useEffect calls, try with the navigation state and lets see if it fails. (it didn't fail when @eh2077 implemented the stack approach using the event listener)

**Comment by DylanDylann** [comment]
- @shubham1206agra @MonilBhavsar Creating a new Onyxkeys to store the ```lastVisitedTimestamp``` works well withou Reassure Performance issue. Can you help recheck the PR.
- @getusha
> I think it is because of the useEffect calls

+ No. The useEffect only calls one time when we open the report
+ And as I mentioned above, we need to prevent updating the invalid report (if using navigation event, we cannot check whether the report is invalid or not).

**Comment by MonilBhavsar** [comment]
As discussed internally, we're little sideways to create a new Onyx container to store this data. So what was the cause for regression tests failure?

**Comment by DylanDylann** [comment]
@MonilBhavsar We call the ```updateLastVisitedTime``` when we know that the report is valid (existed report), so other components that subscribe to the report collection will be re-rendered

**Comment by shubham1206agra** [comment]
> @DylanDylann Can you try to use a new collection for this in another PR just to confirm the performance hit?

@DylanDylann I said to create another with this to see the performance.
You can profile the rendering process with react devtools.
Can you do that with original change? And make the current change in seperate PR just in case everything fails.

**Comment by DylanDylann** [comment]
@shubham1206agra 
https://github.com/Expensify/App/pull/32468: Original solution with the Performance Regression Tests failed
https://github.com/Expensify/App/pull/31424: The new solution using the new Onyxkey

**Comment by MonilBhavsar** [comment]
>  when we know that the report is valid (existed report), so other components that subscribe to the report collection will be re-rendered

I see, we considered re-rendering being one of the cons of this approach. But thought it would be fixable.

**Comment by shubham1206agra** [comment]
> > when we know that the report is valid (existed report), so other components that subscribe to the report collection will be re-rendered
> 
> I see, we considered re-rendering being one of the cons of this approach. But thought it would be fixable.

Thats why I thought to pass this update along with optimistic API call of `OpenReport`, so that extra re-render won't happen.

**Comment by getusha** [comment]
>https://github.com/Expensify/App/pull/31424: The new solution using the new Onyxkey

@DylanDylann that looks similar to the approach we avoided https://github.com/Expensify/App/pull/31097, we should not create a new collection to store time stamps, that's worse than the stack approach, let's just use the event from stack navigator.

**Comment by DylanDylann** [comment]
@shubham1206agra Using your suggestion still encounters the perf regression https://github.com/Expensify/App/pull/32540
@MonilBhavsar What do you think about store the ```lastVisitTime``` in REPORT_METADATA ?

**Comment by getusha** [comment]
>Please help check this #31424 (comment)

@DylanDylann i don't think that'll be an issue, as it was working perfectly on the stack approach PR

**Comment by getusha** [comment]
>@MonilBhavsar What do you think about storing the lastVisitTime in REPORT_METADATA ?

Please check this https://github.com/Expensify/App/issues/30778#issuecomment-1803571077

**Comment by DylanDylann** [comment]
@getusha Update the ```lastVisitTime``` with the invalid reportID can lead to the bellow error:
image

**Comment by DylanDylann** [comment]
@getusha  Based on this [comment](https://github.com/Expensify/App/issues/30778#issuecomment-1803571077)

> So we will anyway need to introduce a new key.

So why don't we create a new key for this PR?

**Comment by DylanDylann** [comment]
@techievivek What do you think about [this](https://github.com/Expensify/App/pull/31424#issuecomment-1842455437)?

**Comment by DylanDylann** [comment]
@MonilBhavsar Why dont we want to create a new Onyxkeys for this?

**Comment by shubham1206agra** [comment]
OK, I think a proper solution would be to use onyx selectors to exclude the lastVisitTime from causing re-rendering in the components that are causing problems. Was that considered?

@DylanDylann Please try this

**Comment by getusha** [comment]
>But it cannot fix the issue.

@DylanDylann isn't the performance failing only on `ReportActionList` component?

**Comment by DylanDylann** [comment]
Not a new collection, just use the REPORT_METADATA, that use `ve used to improve the performance

**Comment by DylanDylann** [comment]
@MonilBhavsar Please help check this comment https://github.com/Expensify/App/pull/31424#issuecomment-1842107877

**Comment by MonilBhavsar** [comment]
Have we tried all the possible ways to fix this? If so, we can store it in report metadata collection. Let's also hear from @ospfranco who added that collection for performance optimization

**Comment by DylanDylann** [comment]
@ospfranco More context for you:
- Once we open the report, we need to store the ```lastVisitTime: new Date()``` for that report.
- With the current solution, we are storing the ```lastVisitTime``` in the ONYXKEYS.COLLECTION.REPORT${reportID} but it will leads to the perf regression.
- So I plan to store the ```lastVisitTime``` in  ONYXKEYS.COLLECTION.REPORT_METADATA${reportID}

**Comment by shubham1206agra** [comment]
We may not get any reply here as @ospfranco is no longer in margelo. So let me recheck the current solution, and we should proceed with that only.

**Comment by shubham1206agra** [comment]
@DylanDylann I am unable to load webpage. It gets stuck on skeleton for me

**Comment by DylanDylann** [comment]
@shubham1206agra Can you share the screenshot? I cannot reproduce

**Comment by shubham1206agra** [comment]
@DylanDylann Can you resolve conflicts quickly?

**Comment by DylanDylann** [comment]
@shubham1206agra I just fixed

**Comment by shubham1206agra** [comment]
@DylanDylann Steps to repro the error
1. Log out
2. Log in with different account
3. See the skeleton infinitely

File: src/libs/ReportUtils.ts [hunk 9]
```diff
@@ -550,12 +550,14 @@ function isDraftExpenseReport(report: OnyxEntry<Report>): boolean {
...
     return Object.values(reports ?? {})
-        .filter((report) => !!report?.reportID && !!report?.lastReadTime)
+        .filter(
+            (report) => !!report?.reportID && ((reportsMetadata && !!reportsMetadata[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${report.reportID}`]?.lastVisitTime) ?? !!report?.lastReadTime),
+        )
         .sort((a, b) => {
-            const aTime = new Date(a?.lastReadTime ?? '');
-            const bTime = new Date(b?.lastReadTime ?? '');
+            const aTime = new Date((reportsMetadata && a && reportsMetadata[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${a.reportID}`]?.lastVisitTime) ?? a?.lastReadTime ?? '');
+            const bTime = new Date((reportsMetadata && b && reportsMetadata[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${b.reportID}`]?.lastVisitTime) ?? b?.lastReadTime ?? '');
```
Comment (shubham1206agra) [inline src/libs/ReportUtils.ts:560]: This seems wrong
```suggestion
            const aTime = new Date(reportsMetadata?.[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${a?.reportID}`]?.lastVisitTime ?? a?.lastReadTime ?? '');
            const bTime = new Date(reportsMetadata?.[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${b?.reportID}`]?.lastVisitTime ?? b?.lastReadTime ?? '');
```

File: src/libs/ReportUtils.ts [hunk 10]
```diff
@@ -550,12 +550,14 @@ function isDraftExpenseReport(report: OnyxEntry<Report>): boolean {
 /**
  * Given a collection of reports returns them sorted by last read
  */
-function sortReportsByLastRead(reports: OnyxCollection<Report>): Array<OnyxEntry<Report>> {
+function sortReportsByLastRead(reports: OnyxCollection<Report>, reportsMetadata: OnyxCollection<ReportMetadata>): Array<OnyxEntry<Report>> {
     return Object.values(reports ?? {})
-        .filter((report) => !!report?.reportID && !!report?.lastReadTime)
+        .filter(
+            (report) => !!report?.reportID && ((reportsMetadata && !!reportsMetadata[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${report.reportID}`]?.lastVisitTime) ?? !!report?.lastReadTime),
+        )
```
Comment (shubham1206agra) [inline src/libs/ReportUtils.ts:557]:
```suggestion
            (report) => !!report?.reportID && !!(reportsMetadata?.[`${ONYXKEYS.COLLECTION.REPORT_METADATA}${report.reportID}`]?.lastVisitTime ?? report?.lastReadTime)
        )
```

**Comment by shubham1206agra** [comment]
@MonilBhavsar Now, you can proceed with final review.
I have rechecked and it works fine now

File: src/libs/ReportUtils.ts [hunk 11]
```diff
@@ -550,12 +550,12 @@ function isDraftExpenseReport(report: OnyxEntry<Report>): boolean {
 /**
  * Given a collection of reports returns them sorted by last read
  */
-function sortReportsByLastRead(reports: OnyxCollection<Report>): Array<OnyxEntry<Report>> {
+function sortReportsByLastRead(reports: OnyxCollection<Report>, reportsMetadata: OnyxCollection<ReportMetadata>): Array<OnyxEntry<Report>> {
```
Comment (shubham1206agra) [inline src/libs/ReportUtils.ts:553]: @DylanDylann Instead of passing reportsMetadata as a parameter, can we use global variable in ReportUtils and use that?

Comment (DylanDylann) [reply src/libs/ReportUtils.ts:553]: Because the we are passing the reports as a parameter, so i think we should do it with reportsMetadata as well for consitency

File: src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts [hunk 12]
```diff
@@ -16,6 +16,8 @@ type ReportScreenIDSetterComponentProps = {
 
     /** Whether user is a new user */
     isFirstTimeNewExpensifyUser: OnyxEntry<boolean>;
+
+    reportsMetadata: OnyxCollection<ReportMetadata>;
```
Comment (MonilBhavsar) [inline src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts:20]:
```suggestion
    reportMetadata: OnyxCollection<ReportMetadata>;
```

File: src/types/onyx/Report.ts [hunk 13]
```diff
@@ -44,7 +44,7 @@ type Report = {
     /** The time of the last read of the report */
     lastReadCreated?: string;
 
-    /** The last time the report was visited */
+    /** The time when user read the last message */
```
Comment (MonilBhavsar) [inline src/types/onyx/Report.ts:47]: thanks!

**Review by MonilBhavsar — COMMENTED** [review]
Minor comments, else looks good!

**Comment by DylanDylann** [comment]
@MonilBhavsar I just updated the PR

**Comment by MonilBhavsar** [comment]
@DylanDylann the PR author checklist has been updated and you also need to update it

**Comment by DylanDylann** [comment]
@MonilBhavsar I updated

File: src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts [hunk 14]
```diff
@@ -16,6 +16,8 @@ type ReportScreenIDSetterComponentProps = {
 
     /** Whether user is a new user */
     isFirstTimeNewExpensifyUser: OnyxEntry<boolean>;
+
+    reportMetadata: OnyxCollection<ReportMetadata>;
```
Comment (MonilBhavsar) [inline src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts:20]: Oops missed this, but props missing a comment.
A part of checklist 🙈

Comment (MonilBhavsar) [reply src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts:20]: >  All props are defined accurately and each prop has a /** comment above it */

Comment (DylanDylann) [reply src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts:21]: Updated @MonilBhavsar

**Comment by MonilBhavsar** [comment]
PR Reviewer checklist also needs to be updated now cc @shubham1206agra

File: src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts (hunk 14 above)
Comment (MonilBhavsar) [reply src/libs/Navigation/AppNavigator/ReportScreenIDSetter.ts:21]: Thanks!

**Comment by shubham1206agra** [comment]
Done

**Review by MonilBhavsar — APPROVED** [review]
Thanks all 🚀

File: src/libs/actions/Report.ts [hunk 15]
```diff
@@ -499,6 +499,7 @@ function openReport(
                 isLoadingInitialReportActions: true,
                 isLoadingOlderReportActions: false,
                 isLoadingNewerReportActions: false,
+                lastVisitTime: DateUtils.getDBTime(),
```
[the remark this answers is not in the document]
Comment (srikarparsi) [reply src/libs/actions/Report.ts:502]: Hi @MonilBhavsar! Just wanted to bump this question in case you know the answer or if you think anyone above would know the answer.

[the remark this answers is not in the document]
Comment (MonilBhavsar) [reply src/libs/actions/Report.ts:502]: This proposal explains it https://github.com/Expensify/App/issues/30868#issuecomment-1793304559
tldr; If user visits a chat and refereshes it, we want to display last visited chat and not some random chat. Last visited chat is determined by `report.lastVisitTime`. So whenever user opens the report, we update that property.

[the remark this answers is not in the document]
Comment (MonilBhavsar) [reply src/libs/actions/Report.ts:502]: May be for somecase where that hook in ReportScreen is not trigerred. @DylanDylann could you please answer this

[the remark this answers is not in the document]
Comment (DylanDylann) [reply src/libs/actions/Report.ts:502]: @parasharrajat @MonilBhavsar 

When I created this PR, setting `lastVisitTime: DateUtils.getDBTime()` in the `openReport` function did not cause any side effects. It was added to ensure that lastVisitTime is always updated whenever a report is opened but I also find any issue if we remove that line.

I just reviewed the code, I found that calling `Report.updateLastVisitTime(reportID)` alone is sufficient. Therefore, we can safely remove the `lastVisitTime: DateUtils.getDBTime()` parameter from openReport.
